### DISCUSSION TITLE
Using middleware to create "templates" for different paths

### DISCUSSION THREAD

**Discussion by dattaproffs** [honojs/hono#2245] [Q&A]
Hi 
i'm trying to setup a structure with middleware to support both full reload and htmx frontend using cloudflare workers.
It is working pretty good but i have som troubles understanding some things.

I have looked at https://github.com/dctanner/htmljs-todo-example wich is pretty much what i wan't to do.
Having trouble to getting redirect to work and some other set / get issues. 
I will post some pseudo code below so maybe someone can tell me what i'm doing wrong and that a much better way exists :) 

```
This should only happen if we are not in a htmx env, eg. hx-request header exists
app.use('*',  SITE LAYOUT... ) eg. <html><body>{children}</body></html> 

Same as above, this should only happen if full request, eg not when hx-request header exists
app.use('/auth/*', SECTION LAYOUT..) eg. <div class="w-full"><h1>AUTH</h1>{children}</div>

app.get('/auth/login', VIEW )) eg. 
export const LoginRoute = async ({ context }: { context: Context<Env, string, {}> }) => {
  return (
    <>
      <div class="sm:mx-auto sm:w-full sm:max-w-sm">
        <h2 class="mt-10 text-center text-2xl font-bold leading-9 tracking-tight text-white">Sign in to your account</h2>
      </div>
        <LoginForm />
      </div>
    </>
  )
}

```

and so on... 
The problems i'm facing is that in for example `app.post('/auth/login', VIEW )`  i wan't to redirect via htmx if in that context otherwise i wan't to do a context.redirec().
But when i try that it does not happen and i think it's because the several middlewares?

The second thing i have a hard time getting to work properly is to get everything typed. I have looked through issues and docs but can't seem to find anything that looks like this. 

If someone could steer me in the right direction i would be greatful.

Thank's in advance and for a great project!

**Comment by yusukebe** [comment]
Hi @dattaproffs 

I don't understand it all, but have you ever tried to use JSX Renderer middleware?

https://hono.dev/middleware/builtin/jsx-renderer

**Reply by dattaproffs** [reply]
hi again @yusukebe 
here  is a simple pseudo code test https://github.com/dattaproffs/honotest1

i have created three different jsxRenderers `rootRenderer` `mainRenderer` `adminMainRenderer`
if you start it up and klick through home, page1, page2 (just redirects to page1) and login you will see that htmx kicks in and we do not render. So far so good, it also works perfectly fine without javascript enabled. 

Now if yo go to login and use admin or admin2 as username you will get "logged in" 
here is where i struggle, i wan't to render a new layout for this. basicly i want to redirect to `/admin` and use `rootRenderer` and `adminMainRenderer` and from there use the same logic as above.

don't think about the code looking messy i have just added a couple of different ways to do stuff. 
As before i have this basicly working i a couple of different ways but none of them seems intuitive and something to build on.

If you have any ideas that can put me on the right track it would be much appreciated

Thanks

**Reply by dattaproffs** [reply]
ok i'm starting to get something working here. Will try to update and create a simple example site when i have time. But seems lite jsxRenderer is the way when trying it out more, thanks will close this.

**Comment by dattaproffs** [comment]
Hi @yusukebe thank's for your reply. Yes i started looking at it yesterday and going to setup a minimal example that we can look at. 
bu basicly i want to conditionally render the "parent" layout depending on if we are in js world (eg. htmx) 

so either i will redirect by setting the htmx-location header or the context.redirect and so on. 
when htmx calls for example POST AddStuff i wan't the response to be only what i have in that method but if JS is disabled i want the whole tree to render. 

Makes some sense? 

I will get back with a small example.
